nab/sweeper.py

```diff
@@ -0,0 +1,245 @@
+# ----------------------------------------------------------------------
+# Copyright (C) 2014-2015, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+from collections import namedtuple
+import logging
+import math
+
+from nab.corpus import Corpus
+
+logger = logging.getLogger(__name__)
+AnomalyPoint = namedtuple("AnomalyPoint", ["timestamp", "anomalyScore", "sweepScore", "windowName"])
+ThresholdScore = namedtuple("ThresholdScore", ["threshold", "score", "tp", "tn", "fp", "fn", "total"])
+
+
+def sigmoid(x):
+  """Standard sigmoid function."""
+  return 1 / (1 + math.exp(-x))
+
+
+def scaledSigmoid(relativePositionInWindow):
+  """Return a scaled sigmoid function given a relative position within a
+  labeled window.  The function is computed as follows:
+
+  A relative position of -1.0 is the far left edge of the anomaly window and
+  corresponds to S = 2*sigmoid(5) - 1.0 = 0.98661.  This is the earliest to be
+  counted as a true positive.
+
+  A relative position of -0.5 is halfway into the anomaly window and
+  corresponds to S = 2*sigmoid(0.5*5) - 1.0 = 0.84828.
+
+  A relative position of 0.0 consists of the right edge of the window and
+  corresponds to S = 2*sigmoid(0) - 1 = 0.0.
+
+  Relative positions > 0 correspond to false positives increasingly far away
+  from the right edge of the window. A relative position of 1.0 is past the
+  right  edge of the  window and corresponds to a score of 2*sigmoid(-5) - 1.0 =
+  -0.98661.
+
+  @param  relativePositionInWindow (float)  A relative position
+                                            within a window calculated per the
+                                            rules above.
+
+  @return (float)
+  """
+  if relativePositionInWindow > 3.0:
+    # FP well behind window
+    return -1.0
+  else:
+    return 2*sigmoid(-5*relativePositionInWindow) - 1.0
+
+
+
+class Sweeper(object):
```

Comment:
Add class level comment with description of what a sweeper is or what it is for. What is it mean to sweep and why? etc.

Reply:
Class-level comment added.